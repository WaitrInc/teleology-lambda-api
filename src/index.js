export { default as wrapper } from './wrapper';
export { default as request } from './request';
