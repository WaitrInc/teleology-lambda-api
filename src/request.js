import qs from 'querystring';

const extractBody = ({ body, isBase64Encoded }) => {
  try {
    const debuff = Buffer.from(body, isBase64Encoded ? 'base64' : undefined);
    const o = JSON.parse(debuff);
    if (o && typeof o === 'object') {
      return o;
    }
  } catch (e) {
    // do nothing
  }

  return body;
};

export default ({
  headers,
  pathParameters,
  queryStringParameters,
  ...event
}) => {
  const h = Object.keys(headers).reduce((a, key) => {
    a[key.toLowerCase()] = headers[key];
    return a;
  }, {});

  return {
    ...event,
    headers: h,
    data: {
      ...pathParameters,
      ...queryStringParameters,
      ...extractBody(event),
      ...qs.parse(event.body || ''),
    },
  };
};
