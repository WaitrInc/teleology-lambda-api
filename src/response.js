const DEFAULT_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Credentials': true,
  'Content-Type': 'application/json',
};

export default ({ code = 200, body = '', headers }) => ({
  statusCode: `${code}`,
  body: typeof body === 'string' ? body : JSON.stringify(body),
  headers: {
    ...DEFAULT_HEADERS,
    ...headers,
  },
});
