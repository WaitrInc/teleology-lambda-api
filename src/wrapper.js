/* eslint-disable no-param-reassign */
import request from './request';
import response from './response';

export default (fn) => async (event) => {
  try {
    const { body, headers = {}, ...rest } = await fn(request(event));
    return response({
      code: 200,
      headers,
      body: body || rest,
    });
  } catch (e) {
    // todo: coming soon: a way to configure this that doesn't rely on serverless-offline
    if (process.env.IS_OFFLINE === "true") { /* true if using serverless-offline */
      // AWS framework will handle this
      throw e;
    } else {
      return new Promise((_, reject) => {reject(e.toString())});
    }
  }
};
