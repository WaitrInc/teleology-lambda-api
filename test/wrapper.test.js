import { wrapper, ApiError } from '../src';

const event = {
  pathParameters: {
    slug: 'test',
  },
  queryStringParameters: {
    next: 10,
  },
};

describe.only('wrapper', () => {
  it('success defaults', async () => {
    const handler = () => ({
      foo: 'bar',
    });

    const result = await wrapper(handler)(event);
    expect(result).toMatchSnapshot();
  });

  it('success custom headers and body', async () => {
    const handler = () => ({
      headers: {
        'Content-Type': 'text/html',
      },
      body: '<h1>hello</h1>',
    });

    const result = await wrapper(handler)(event);
    expect(result).toMatchSnapshot();
  });

  it('error defaults', async () => {
    const error = new Error('foo');

    const handler = () => {
      throw error;
    };

    await expect(wrapper(handler)(event)).rejects.toEqual(error.toString());
  });

  it('custom error', async () => {
    const error = new ApiError('Unauthorized', { code: 401, description: 'nope' });

    const handler = () => {
      throw error;
    };

    await expect(wrapper(handler)(event)).rejects.toEqual(error.toString());  
  });

  it('custom error, when IS_OFFLINE, rethrow', async () => {
    const msg = 'Unauthorized';
    const error = new ApiError(msg, { code: 401, description: 'nope' });
    // move to before/after if there are any more tests that require this env variable
    const envBefore = process.env.IS_OFFLINE;
    try {
      process.env.IS_OFFLINE = "true";

      const handler = () => {
        throw error;
      };

      await expect(wrapper(handler)(event)).rejects.toThrow(msg);  
    } catch(e) {
      process.env.IS_OFFLINE = envBefore;
      throw e;
    }
  });
});
