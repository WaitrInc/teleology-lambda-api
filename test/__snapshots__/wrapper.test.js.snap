// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapper success custom headers and body 1`] = `
Object {
  "body": "<h1>hello</h1>",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "text/html",
  },
  "statusCode": "200",
}
`;

exports[`wrapper success defaults 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": "200",
}
`;
